### PR TITLE
Remove .source files for PG15 support

### PR DESCRIPTION
**Pull request by hanefi**
PostgreSQL 15 dropped usage of .source files that are used to generate
.sql and .out files by replacing some placeholders with the actual
values before test runs. Instead, the information is passed from
pg_regress to the .sql and .out files directly via env variables. Those
variables are read via \getenv psql command in relevant test files.

PostgreSQL 15 introduced some changes to pg_regress binary that allowed
this to happen. However this change is not backported to earlier
versions, and thus we come up with a similar mechanism in
pg_regress_multi that works in all supported PG versions.

We also needed to make some changes to `copy` and `\copy` commands.
`\copy` does not support variable interpolation, and we need to store
`\copy` commands in a variable and use that variable in a consecutive
line to let interpolation do its magic.

Relevant PG commits:
- postgres/postgres@33d3eeadb21d2268104840cfef6bc2226ddfc680
  adds `\getenv` command to psql.
- postgres/postgres@d1029bb5a26cb84b116b0dee4dde312291359f2a
  updates all `.source` files to be supported `sql` or `out` files
  without actually renaming them. `pg_regress.c` is patched to set some
  environment variables that contain paths to relevant directories, and
  the `.source` files use the newly introduced `\getenv` to read those
  paths.
- postgres/postgres@dc9c3b0ff21465fa89d71eecf5e6cc956d647eca
  renames all `.source` files into either `.sql` or `.out` files.

Closes: #6107

### CODE REVIEW

File: src/test/regress/.gitignore [hunk 1]
```diff
@@ -38,7 +38,6 @@ create_function_2.sql
 largeobject.sql
 misc.sql
 security_label.sql
-tablespace.sql
```
Comment (hanefi) [inline src/test/regress/.gitignore:41]: Postgres Vanilla tests also include a test with the same name. Hence we had gitignore rules for this file in 2 locations.

The best solution I can think of is renaming this to `multi_tablespaces` so that we avoid name collisions. If this looks like a good idea, I can open a separate PR that addresses this.

File: ci/README.md [hunk 2]
```diff
@@ -363,11 +363,8 @@ This was deemed to be error prone and not worth the effort.
 
 This script checks and fixes issues with `.gitignore` rules:
 
-1. Makes sure git ignores the `.sql` files and expected output files that are generated
-   from `.source` template files. If you created or deleted a `.source` file in a commit,
-   git ignore rules should be updated to reflect this change.
 
-2. Makes sure we do not commit any generated files that should be ignored. If there is an
+1. Makes sure we do not commit any generated files that should be ignored. If there is an
```
Comment (hanefi) [inline ci/README.md:367]: Does it make sense to have a numbered list if we have only one item left? I am not so sure.

**Comment by hanefi** [comment]
Notes to reviewer:
- I tried to split my work into meaningful commits, and it may be easier to review them separately.
- When a file is updated, and later moved in separate commits, GitHub does not always show them as a file move but two independent files. In that case it is easier to review commits separately. 
- I intentionally changed most `\copy` commands into `COPY` sql queries because the former does not support variable interpolation. Hence we need store and execute `\copy` commands that use data paths in two steps for interpolation to kick in.
- I verified that the tests work as intended on PG15 as well. Hence the prototype PR that should be closed after this one is merged. https://github.com/citusdata/citus/pull/6107

File: ci/README.md (hunk 2 above)
Comment (onderkalaci) [reply ci/README.md:367]: I'm fine either way

File: src/test/regress/expected/multi_agg_distinct.out [hunk 3]
```diff
@@ -154,61 +156,61 @@ CREATE TABLE lineitem_hash (
...
---------------------------
- 
+ create_distributed_table
+---------------------------------------------------------------------
+
 (1 row)
 
-\copy lineitem_hash FROM '@abs_srcdir@/data/lineitem.1.data' with delimiter '|'
-\copy lineitem_hash FROM '@abs_srcdir@/data/lineitem.2.data' with delimiter '|'
+COPY lineitem_hash FROM :'lineitem_1_data_file' with delimiter '|';
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_agg_distinct.out:164]: Do we have any `\copy` tests remaining in the PR ? I guess you can still do `\copy` from program, and it seems useful to have at least one such test on distributed tables?

File: src/test/regress/expected/multi_outer_join.out [hunk 4]
```diff
@@ -877,15 +885,15 @@ SELECT create_distributed_table('left_values', 'val');
 
 (1 row)
 
-\copy left_values from stdin
+COPY left_values from stdin;
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_outer_join.out:888]: Not important if that is just copy&paste error, but is there a reason we needed to change this?

**Review by onderkalaci — COMMENTED** [review]
I think the PR itself looks fine, with minor questions.

Though, locally it fails on `check-multi`, can you check why? I'm using pg14

```
waiting for server to start.... done
server started
waiting for server to start.... done
server started
waiting for server to start.... done
server started
CREATE DATABASE
CREATE EXTENSION
CREATE FUNCTION
CREATE FOREIGN DATA WRAPPER
CREATE DATABASE
CREATE EXTENSION
CREATE FUNCTION
CREATE FOREIGN DATA WRAPPER
(using postmaster on localhost, port 57636)
pg_regress: no *.source files found in "/Users/onderkalaci/Documents/citus_code/citus/src/test/regress/input"
mv: rename results/* to results/*.unmodified: No such file or directory
waiting for server to shut down.... done
server stopped
waiting for server to shut down.... done
server stopped
waiting for server to shut down.... done
server stopped
Failed in 0 seconds. 
make: *** [check-multi] Error 25
```

**Comment by thanodnl** [comment]
> * I intentionally changed most `\copy` commands into `COPY` sql queries because the former does not support variable interpolation. Hence we need store and execute `\copy` commands that use data paths in two steps for interpolation to kick in.

Will this cause problems with running our tests against Hyperscale?

There is a difference between `\copy` and `COPY` where the files are read, `\copy` reads them in psql iirc, while `COPY` reads them on the server.

Not to say that we definitely should keep that behaviour, and I don't know how often we run our regression tests against hyperscale, but it is something to keep in mind.

File: src/test/regress/expected/multi_outer_join.out (hunk 4 above)
Comment (hanefi) [reply src/test/regress/expected/multi_outer_join.out:888]: No reason. This was a regex error. I can revert this change.

I replaced all `\copy` commands into `COPY` because only the latter support variable interpolation. However, when we do not have reference to local files, both can work just fine.

**Comment by onderkalaci** [comment]
> (using postmaster on localhost, port 57636)
pg_regress: no *.source files found in "/Users/onderkalaci/Documents/citus_code/citus/src/test/regress/input"
mv: rename results/* to results/*.unmodified: No such file or directory
waiting for server to shut down.... done

As @hanefi noted in private to me that I got the error because I still had `source` and `output` folders on my local machine. And, after removing those folders, the tests pass.

File: src/test/regress/expected/multi_agg_distinct.out (hunk 3 above)
Comment (hanefi) [reply src/test/regress/expected/multi_agg_distinct.out:164]: Several examples:

`multi_copy`

```sql
-- Test client-side copy from file
-- \copy does not support variable interpolation. Hence we store and execute
-- the query in two steps for interpolation to kick in.
-- See https://stackoverflow.com/a/67642094/4576416 for details.
\set client_side_copy_command '\\copy customer_copy_hash FROM ' :'customer3datafile' ' WITH (DELIMITER '''|''');'
:client_side_copy_command
```

-----

`pg_dump` contains 4 `\copy .. FROM PROGRAM`
```sql
\COPY output FROM PROGRAM 'pg_dump --quote-all-identifiers -h localhost -p 57636 -U postgres -d regression -t dumper.data --data-only | psql -tAX -h localhost -p 57636 -U postgres -d regression'
```

-----

`multi_multiuser_load_data` that checks various user scenarios. `COPY` needs superuser privileges to read from files on disk, and hence we use `\copy` there instead.

```sql
-- as user with no access, should fail
SET ROLE no_access;
\set copy_command '\\COPY lineitem FROM ' :'lineitem_2_data_file' ' with delimiter '''|''';'
:copy_command
RESET ROLE;
```

Comment (hanefi) [reply src/test/regress/expected/multi_agg_distinct.out:164]: After another check, I realized that the only `\copy .. FROM PROGRAM` tests write to local tables. Let me update an existing `COPY` test that uses a PROGRAM clause into `\copy`.

**Comment by hanefi** [comment]
> Will this cause problems with running our tests against Hyperscale?

Yes it will. The tests on https://github.com/citusdata/citus/blob/main/src/test/regress/multi_schedule_hyperscale should not use `COPY` but `\COPY` commands. However I want to move forward with this solution as is because we do not already do a good job maintaining tests against HyperScale.

> Not to say that we definitely should keep that behaviour, and I don't know how often we run our regression tests against hyperscale, but it is something to keep in mind.

I tried to try running the tests on HyperScale with no success. The documentation points to Scout (no longer an option) that uses our old HyperScale Citus infra (that I shall not name here). We need to spend a significant amount of effort to get HyperScale tests in a meaningful state that could be helpful. Some ideas to get the hyperscale schedule tests run and provide some value:
1. Do not turn off MX by default
2. Run the tests on newer HyperScale infra
3. Go over `multi_schedule_hyperscale` and add more test files. IIRC, developers were expected to add new test files in that schedule, but that did not happen even once in the last year.
4. Go over the ideas in https://github.com/citusdata/citus/pull/5042 and fix them.

In short, I believe it is worth keeping shorter and easier to read `COPY` tests instead of `\COPY` ones so that
a) We have fewer lines of tests
b) We do not have messy string manipulation logic for escaping bunch of characters that ends up hurting readability.

**Comment by hanefi** [comment]
> As @hanefi noted in private to me that I got the error because I still had `input` and `output` folders on my local machine. And, after removing those folders, the tests pass.

I updated makefile target for `make clean` on `src/test/regress` directory to remove those directories that should now be empty. This will make it easier for developers to get the tests to run locally.

The culprit is the following:
- Git tracks files, not directories. Even though we removed all the files inside `input` and `output` directories, there is no simple way to remove the empty directories in an old git tree.
- pg_regress on PG<15 assumes that if there are `input` or `output` directory, there is a positive number of `.source` files inside them. This creates issues on old git clones that used to have some `.source` files in them.

**Comment by onderkalaci** [comment]
> I tried to try running the tests on HyperScale with no success. The documentation points to Scout (no longer an option) that uses our old HyperScale Citus infra (that I shall not name here). We need to spend a significant amount of effort to get HyperScale tests in a meaningful state that could be helpful. Some ideas to get the hyperscale schedule tests run and provide some value:

I was thinking whether we can point arbitrary configs to a HSC cluster. That should avoid the extra maintenance of HSC tests.

> However I want to move forward with this solution as is because we do not already do a good job maintaining tests against HyperScale.

Don't we run these tests as part of release testing? It seems we have NOT run this for Citus 11.0 release testing. Maybe @metdos or @halilozanakgul might know the reason.

But, in general, I don't think we should be breaking a whole test suite unless we at least have a way to unblock those. 

> b) We do not have messy string manipulation logic for escaping bunch of characters that ends up hurting readability.

I really wouldn't mind having this. In the end, we barely (or even never) add COPY from file anymore. So, those will probably never be read frequently anyway? I think adding the logic on how to calculate that in the PR description would help a lot

**Review by onderkalaci — APPROVED** [review]

**Review by thanodnl — APPROVED** [review]
lgtm

Thanks for the effort here!
